lib/declare_schema/model/index_definition.rb

```diff
@@ -51,11 +57,20 @@ def for_table(table_name, ignore_indexes, connection)
                 raise "primary key on #{table_name} was not unique on #{primary_key_columns} (was unique=#{index.unique} on #{index.columns})"
               primary_key_found = true
             end
-            new(index.columns, name: index.name, unique: index.unique, where: index.where)
+            length =
+              case lengths = index.lengths
+              when {}
+                nil
+              when Hash
+                lengths.size == 1 ? lengths.values.first : lengths
```

Comment:
❓ What are the different hash shapes that `index.lengths` can take? This contract we're building feels brittle to me, so I'm curious if we can harden it with more strict expectations.

Reply:
It appears to always return a Hash. (Not what I expected at first.) So a fresh cut at the code could be much cleaner:
```ruby
lengths_hash = index.lengths
length =
  case lengths_hash.size
  when 0
    nil
  when 1
    lengths_hash.values.first
  else
    lengths_hash
  end
```
I should probably try some of the corner cases to make sure they line up and don't get stuck in an endless migration loop due to aliasing, though. For example, I believe a single-column partial index can be requested either way in ActiveRecord:
```ruby
add_index :users, [:last_name], length: 10
```
or
```ruby
add_index :users, [:last_name], length: { last_name: 10 }
```
and also I wonder if ActiveRecord will allow the opposite on multi-column partial indexes, where they'll apply one length to all the columns?
```ruby
add_index :users, [:last_name, :first_name], length: 10
```

Reply:
Ah, I see. Thank you. Maybe we should utilize the same contract where `length` will always be a hash as well so that we don't have to have so many cases else where for handing different shapes of the length value. This could simply turn into:

```ruby
length = index.lengths
(length.nil? || length === Hash) or raise "Unexpected value #{length.inspect} provided for index lengths, expected a Hash"
```

Reply:
That's an interesting compromise. I still like the simple Railsy configuration for single-column  indexes:
```ruby
index [:last_ name], length: 10
```
vs requiring this less-DRY form:
```ruby
index [:last_name], length: { last_name: 10 }
```
and of course there's symbol vs String aliasing too. :shakes_fist:

That all leads probably needing a normalize function. Something like
```
def normalize_index_length(columns, value)
  case value
  nil
    nil
  {}
    nil
  Hash
    value.transform_keys(&:to:s)
  Integer
    columns.each_with_object({}) { |column, hash| hash[column.to_s] = value }
  end
end
```
I could run the "desired state" config through that as well as what comes back from SQL in the `indexes` result.